contracts/libraries/AddressPermissionControl.sol

```diff
@@ -0,0 +1,147 @@
+        return _permissionMasks[target];
+    }
+
+    function stagedAddresses() public view returns (address[] memory) {
+        return _stagedAddresses.values();
+    }
+
+    function stagedAddressesLength() public view returns (uint256) {
+        return _stagedAddresses.length();
+    }
+
+    function stagedAddressAt(uint256 index) public view returns (address) {
+        return _stagedAddresses.at(index);
+    }
+
+    function stagedPermissionMaskOf(address target) public view returns (uint256) {
+        return _stagedPermissionMasks[target];
+    }
+
+    function _hasPermission(address addr, uint8 permissionId) internal view returns (bool) {
+        return _permissionMasks[addr] & _permissionIdToMask(permissionId) != 0;
+    }
+
+    function _hasStagedPermission(address addr, uint8 permissionId) internal view returns (bool) {
+        return _stagedPermissionMasks[addr] & _permissionIdToMask(permissionId) != 0;
+    }
+
+    function _isStagedToCommit() private view returns (bool) {
+        return _stagedToCommitAt != 0;
+    }
+
+    function _permissionIdToMask(uint8 permissionId) private pure returns (uint256) {
+        return 1 << (permissionId);
+    }
+
+    function _clearStagedPermissions() private {
+        uint256 length = _stagedAddresses.length();
+        for (uint256 i; i != length; ++i) {
+            address target = _stagedAddresses.at(i);
+            delete _stagedPermissionMasks[target];
+            _stagedAddresses.remove(target);
+        }
+    }
+
+    function _revokePermissionInstant(address from, uint8 permissionId) private {
+        uint256 diff = _permissionIdToMask(permissionId);
+        uint256 currentMask = _permissionMasks[from];
+        require(currentMask & diff != 0, ExceptionsLibrary.INVALID_TARGET);
+        _permissionMasks[from] = currentMask & (~diff);
+        if (_permissionMasks[from] == 0) {
+            delete _permissionMasks[from];
+            _addresses.remove(from);
+        }
+    }
+
+    function _revokePermissionsInstant(address from, uint8[] calldata permissionIds) internal {
+        for (uint256 i; i != permissionIds.length; ++i) {
+            _revokePermissionInstant(from, permissionIds[i]);
+        }
+        emit RevokedPermissionsInstant(msg.sender, from, permissionIds);
+    }
+
+    function _stageGrantPermission(address to, uint8 permissionId) private {
+        require(!_isStagedToCommit(), "Already staged");
+        uint256 diff = _permissionIdToMask(permissionId);
+        if (!_stagedAddresses.contains(to)) {
+            _stagedAddresses.add(to);
+            _stagedPermissionMasks[to] = _permissionMasks[to];
+        }
+        uint256 currentMask = _stagedPermissionMasks[to];
+        require(currentMask & diff == 0, ExceptionsLibrary.INVALID_TARGET);
+        _stagedPermissionMasks[to] = currentMask | diff;
+    }
+
+    function _rollbackStagedPermissions() internal {
+        require(_isStagedToCommit(), ExceptionsLibrary.INVARIANT);
```

Comment:
I think `ExceptionsLibrary.NULL` is better here

Reply:
Why `NULL`? Added `ExceptionsLibrary.INVALID_STATE`, I think it suits more

Reply:
The idea was that stagedChanges == NULL but probably state could also work